giftless/storage/amazon_s3.py

```diff
@@ -0,0 +1,112 @@
+from giftless.storage.exc import ObjectNotFound
+from giftless.util import safe_filename
+
+
+class AmazonS3Storage(StreamingStorage, ExternalStorage):
+    """AWS S3 Blob Storage backend.
+    """
+
+    def __init__(self, bucket_name: str, path_prefix: Optional[str] = None, **_):
+        self.bucket_name = bucket_name
+        self.path_prefix = path_prefix
+        self.s3 = boto3.resource('s3')
+        self.s3_client = boto3.client('s3')
+
+    def get(self, prefix: str, oid: str) -> Iterable[bytes]:
+        if not self.exists(prefix, oid):
+            raise ObjectNotFound()
+        result: Iterable[bytes] = self._s3_object(prefix, oid).get()['Body']
+        return result
+
+    def put(self, prefix: str, oid: str, data_stream: BinaryIO) -> int:
+        completed = []
+
+        def upload_callback(size):
+            completed.append(size)
+
+        bucket = self.s3.Bucket(self.bucket_name)
+        bucket.upload_fileobj(data_stream, self._get_blob_path(prefix, oid), Callback=upload_callback)
+        return sum(completed)
+
+    def exists(self, prefix: str, oid: str) -> bool:
+        try:
+            self.get_size(prefix, oid)
+        except ObjectNotFound:
+            return False
+        return True
+
+    def get_size(self, prefix: str, oid: str) -> int:
+        try:
+            result: int = self._s3_object(prefix, oid).content_length
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                raise ObjectNotFound()
+            else:
+                raise e
+        return result
+
+    def get_upload_action(self, prefix: str, oid: str, size: int, expires_in: int,
+                          extra: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+        params = {
+            'Bucket': self.bucket_name,
+            'Key': self._get_blob_path(prefix, oid)
+        }
+        response = self.s3_client.generate_presigned_url('put_object',
+                                                         Params=params,
+                                                         ExpiresIn=expires_in
+                                                         )
+        return {
+            "actions": {
+                "upload": {
+                    "href": response,
+                    "header": {},
+                    "expires_in": expires_in
+                }
+            }
+        }
+
+    def get_download_action(self, prefix: str, oid: str, size: int, expires_in: int,
+                            extra: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
+
+        raw_filename = extra.get('filename') if extra else oid
+        assert raw_filename
+        filename = safe_filename(raw_filename)
+        params = {
+            'Bucket': self.bucket_name,
+            'Key': self._get_blob_path(prefix, oid),
+            'ResponseContentDisposition': f"attachment;filename={filename}"
+        }
```

Comment:
I think it would be nicer to not set content-disposition unless we need to (that is unless the user specified a custom name). 

Something like:

```suggestion
        params = {'Bucket': self.bucket_name, 'Key': self._get_blob_path(prefix, oid)}
        if extra and 'filename' in extra:
            filename = safe_filename(extra['filename'])
            params['ResponseContentDisposition'] = f'attachment; filename="{filename}"'
```
I think this will also pass type checking. 
I've also added double quotes around the filename in the header, it is not strictly required unless the filename contains a space or a semicolon IIRC. While right now `safe_filename` prevents that, in the future I'd like to make it less strict and then it will be required.

Reply:
That's helpful, thank you. I have manually verified and actually the `'ResponseContentDisposition'` can only be provided if we want a custom filename for downloaded file. Otherwise it defaults to object it `oid`.
Type checks also passed this change. Thank you for fixing the syntax of the header too.
I have changed only the `params` dict init so it aligns better with the rest of the dicts in the class.